controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java

```diff
@@ -36,22 +37,24 @@ protected void maintain() {
                     DeploymentMetrics appMetrics = new DeploymentMetrics(metrics.queriesPerSecond(), metrics.writesPerSecond(),
                                                                          metrics.documentCount(), metrics.queryLatencyMillis(), metrics.writeLatencyMillis());
 
-                    // Avoid locking for a long time, due to slow YAMAS.
                     try (Lock lock = controller().applications().lock(application.id())) {
-                        // Deployment (or application) may have changed (or be gone) now:
-                        controller().applications().get(application.id()).ifPresent(freshApplication -> {
-                            Deployment freshDeployment = freshApplication.deployments().get(deployment.zone());
-                            if (freshDeployment != null)
-                                controller().applications().store(freshApplication.with(freshDeployment.withMetrics(appMetrics)), lock);
-                        });
+
+                        // Deployment or application may have changed (or be gone) now:
+                        application = controller().applications().require(application.id());
```

Comment:
require will throw if it is missing, so I think you want to use get() and check Optional.isPresent

Reply:
Ah, I should have checked that :O